rs/http_utils/src/file_downloader.rs

```diff
@@ -182,16 +183,47 @@ pub fn check_file_hash(path: &Path, expected_sha256_hex: &str) -> FileDownloadRe
     }
 }
 
-/// Extract the contents of a given .tar.gz file into `target_dir`
-pub fn extract_tar_gz_into_dir(tar_gz_path: &Path, target_dir: &Path) -> FileDownloadResult<()> {
-    let map_to_untar_error = |e| FileDownloadError::untar_error(tar_gz_path, e);
+/// Check if the file is in Gzip format by verifying the first 2 bytes
+fn is_gz_file<R: Read>(reader: &mut R) -> io::Result<bool> {
+    let mut buffer = [0; 2];
+    reader.read_exact(&mut buffer)?;
+    Ok(buffer == [0x1f, 0x8b])
+}
+
+/// Check if the file is in Zstandard format by verifying the first 4 bytes
+fn is_zst_file<R: Read>(reader: &mut R) -> io::Result<bool> {
+    let mut buffer = [0; 4];
+    reader.read_exact(&mut buffer)?;
+    Ok(buffer == [0x28, 0xb5, 0x2f, 0xfd])
+}
 
-    let tar_gz_file =
-        File::open(tar_gz_path).map_err(|e| FileDownloadError::file_open_error(tar_gz_path, e))?;
+/// Extract the contents of a given .tar.gz or tar.zst file into `target_dir`
+pub fn extract_tar_into_dir(tar_path: &Path, target_dir: &Path) -> FileDownloadResult<()> {
+    let map_to_untar_error = |e| FileDownloadError::untar_error(tar_path, e);
+
+    let tar_file =
+        File::open(tar_path).map_err(|e| FileDownloadError::file_open_error(tar_path, e))?;
+    let mut buf_reader = io::BufReader::new(tar_file);
+
+    if is_gz_file(&mut buf_reader).unwrap_or(false) {
+        let tar_gz_file =
+            File::open(tar_path).map_err(|e| FileDownloadError::file_open_error(tar_path, e))?;
+        let tar = GzDecoder::new(tar_gz_file);
+        let mut archive = Archive::new(tar);
+        archive.unpack(target_dir).map_err(map_to_untar_error)?;
+    } else if is_zst_file(&mut buf_reader).unwrap_or(false) {
```

Comment:
I'm not sure how `BufReader` works exactly, but doesn't the stream position change when we call `is_gz_file` meaning that when we call `is_zst_file` after calling `is_gz_file` we will not read the very first 4 bytes of the file but instead the bytes at positions `2,3,4,5` in the file?

Reply:
+1 

I guess it might also be good to add a test along with the change to make sure the new functionality works as expected.

Reply:
BufReader bug fixed and tests added! Thank you!